src/dynamic_boundary_conditions/tide_data.py

```diff
@@ -1,148 +1,449 @@
+        The start date for data collection, which can be in the past or present.
+    total_days: int = 365
+        The number of days of tide data to collect. The default value is 365 for one year.
+    days_per_call : int = 31
+        The number of days that can be retrieved in a single API call (between 1 and 31 inclusive).
+        The default value is configured to 31, which is the maximum number of days that can be retrieved per API call.
+    """
+    # Check for invalid arguments
+    if total_days < 1:
+        raise ValueError(f"total_days is {total_days}, must be at least 1.")
+    if not 1 <= days_per_call <= 31:
+        raise ValueError(f"days_per_call is {days_per_call}, must be between 1 and 31 inclusive.")
+    # end date for data retrieval
+    end_date = start_date + timedelta(days=total_days - 1)
+    # Initialize an empty dictionary to store the date ranges
+    date_ranges = {}
+    # Loop through the date range, adding each chunk of data to the dictionary
+    while start_date <= end_date:
+        # Determine the start date and number of days for the current chunk
+        request_end_date = min(end_date, start_date + timedelta(days=days_per_call - 1))
+        number_of_days = (request_end_date - start_date).days + 1
+        # Add the current chunk to the date_ranges dictionary
+        date_ranges[start_date] = number_of_days
+        # Move the start date forward for the next chunk
+        start_date += timedelta(days=number_of_days)
+    return date_ranges
+
+
+def gen_api_query_param_list(
+        api_key: str,
+        lat: Union[int, float],
+        long: Union[int, float],
+        datum: DatumType,
+        date_ranges: Dict[date, int],
+        interval: Optional[int] = None) -> List[Dict[str, Union[str, int]]]:
+    """
+    Generate a list of api query parameters used to retrieve high and low tide data for the entire requested period.
+
+    Parameters
+    ----------
+    api_key : str
+        NIWA api key (https://developer.niwa.co.nz/).
+    lat : Union[int, float]
+        Latitude range -29 to -53 (- eg: -30.876).
+    long : Union[int, float]
+        Longitude range 160 to 180 and -175 to -180 (- eg: -175.543).
+    datum : DatumType
+        Datum used. LAT: Lowest astronomical tide; MSL: Mean sea level.
+    date_ranges : Dict[date, int]
+        Dictionary of start date and number of days for each API call that need to be made to retrieve
+        high and low tide data for the entire requested period.
+    interval: Optional[int] = None
+        Output time interval in minutes, range from 10 to 1440 minutes (1 day).
+        Omit to get only high and low tide times.
+    """
+    # Check for invalid arguments
+    if not (-53 <= lat <= -29):
+        raise ValueError(f"latitude is {lat}, must range from -29 to -53.")
+    if not ((160 <= long <= 180) or (-180 <= long <= -175)):
+        raise ValueError(f"longitude is {long}, must range from 160 to 180 and from -175 to -180.")
```

Comment:
```suggestion
    if not ((160 <= long <= 180) or (-180 <= long <= -175)):
        raise ValueError(f"longitude is {long}, must range from 160 to 180 or from -175 to -180.")
```

Reply:
Is there a reason for this? Maybe add some context for these limits eg bbox of nz, As well as for the others

Reply:
amended. just following the Tide API query parameter requirements.